sn_transfers/src/wallet/local_store.rs

```diff
@@ -354,36 +354,41 @@ impl LocalWallet {
         )?;
 
         // cache transfer payments in the wallet
+        let mut cashnotes_to_use: HashSet<CashNote> = offline_transfer
+            .created_cash_notes
+            .iter()
+            .cloned()
+            .collect();
         for (xorname, recipients_info) in recipients_by_xor {
             let (storage_payee, royalties_payee) = recipients_info;
             let node_key = storage_payee.1;
-            let cash_note_for_node = offline_transfer
-                .created_cash_notes
+            let pay_amount = storage_payee.0;
+            let cash_note_for_node = cashnotes_to_use
                 .iter()
                 .find(|cash_note| {
-                    // cash_note.reason() == Hash::hash(xorname) && // NB TODO uncomment and use Reason!!
-                    cash_note.main_pubkey() == &node_key
+                    cash_note.value() == Ok(pay_amount) && cash_note.main_pubkey() == &node_key
```

Comment:
I am not quite sure.
according to your description, the issue is:
when there is a node got paid more than once within the same batch, the current lookup might get it mixed up.

However, as it is targeting the same node, the `storecost pay_amount` quoted will be the same as well in high chance.
Hence, checking the `payment_amount` only won't tell the difference ?

I think the cashnote shall contain the `reason, i.e. which chunk it is paying for`, to make it unqiue.
However, not sure whether it is current provided by cashnote.
If not, shall we introduce it?

Reply:
I agree using the reason would be ideal!As we seemingly don't use Reason in the code, I used the price. 
If the price is identical then the node can't tell the difference as the payment is for them anyways so it works. 

I tried looking into re-establishing reasons. But turns out it's a deeper hole than I thought. 
See, reasons are attached to a `Transaction`. When we batch payments, we create one single transaction for all the chunks so we can't have one reason for all. 
- One option is to use a merkle tree like we had. The issue is that we have to send/store it along as proof. This is reason enough not to do so. 
- Another option which is better is to make one Transaction per chunk. 

That second option has qualities:
- generated CashNotes will be much smaller!
- uploaded Spends will be much smaller!
- each record will have its payment Tx + reason

And drawbacks:
- we have one change CashNote per Tx, so one per chunk
- parallelisation will suffer greatly as if I use one CashNote in a Tx I can't use it for another Tx, I'll have to wait to use the change CashNote once it's valid (first Tx's Spends uploaded). 

I think it's a great feature that we could implement, but later, when we've adapted the way we parallelise.

Reply:
right, yeah I understand it could have big affects to using `reason` for different chunk, to differentiate cash_notes.
But, it could be reason that we have certain `looping of spend got rejected`, and we may have to impl it or some simiar.

I am OK with using `price` as a quick fix.
BUT, I do think this issue shall be raised and get fully discussed and resolved befor next public testnet.